test/jdk/java/util/HashMap/WhiteBoxResizeTest.java

```diff
@@ -21,113 +22,588 @@
-    final VarHandle TABLE;
+    final VarHandle HM_TABLE;
+    final VarHandle WHM_TABLE;
 
     public WhiteBoxResizeTest() throws ReflectiveOperationException {
-        Class<?> mClass = HashMap.class;
-        String nodeClassName = mClass.getName() + "$Node";
-        Class<?> nodeArrayClass = Class.forName("[L" + nodeClassName + ";");
-        MethodHandles.Lookup lookup = MethodHandles.privateLookupIn(mClass, MethodHandles.lookup());
-        TABLE = lookup.findVarHandle(mClass, "table", nodeArrayClass);
-        this.TABLE_SIZE_FOR = lookup.findStatic(
-                mClass, "tableSizeFor",
-                MethodType.methodType(int.class, int.class));
-        this.THRESHOLD = lookup.findVarHandle(mClass, "threshold", int.class);
+        MethodHandles.Lookup hmlookup = MethodHandles.privateLookupIn(HashMap.class, MethodHandles.lookup());
+        TABLE_SIZE_FOR = hmlookup.findStatic(
+                HashMap.class, "tableSizeFor", MethodType.methodType(int.class, int.class));
+        HM_TABLE = hmlookup.unreflectVarHandle(HashMap.class.getDeclaredField("table"));
+
+        MethodHandles.Lookup whmlookup = MethodHandles.privateLookupIn(WeakHashMap.class, MethodHandles.lookup());
+        WHM_TABLE = whmlookup.unreflectVarHandle(WeakHashMap.class.getDeclaredField("table"));
     }
 
+    /*
+     * utility methods
+     */
+
     int tableSizeFor(int n) {
         try {
             return (int) TABLE_SIZE_FOR.invoke(n);
-        } catch (Throwable t) { throw new AssertionError(t); }
+        } catch (Throwable t) {
+            throw new AssertionError(t);
+        }
     }
 
-    Object[] table(HashMap map) {
+    Object[] table(Map<?, ?> map) {
         try {
-            return (Object[]) TABLE.get(map);
-        } catch (Throwable t) { throw new AssertionError(t); }
+            VarHandle vh = map instanceof WeakHashMap ? WHM_TABLE : HM_TABLE;
+            return (Object[]) vh.get(map);
+        } catch (Throwable t) {
+            throw new AssertionError(t);
+        }
     }
 
-    int capacity(HashMap map) {
+    int capacity(Map<?, ?> map) {
         return table(map).length;
     }
 
-    @Test
-    public void testTableSizeFor() {
-        assertEquals(tableSizeFor(0), 1);
-        assertEquals(tableSizeFor(1), 1);
-        assertEquals(tableSizeFor(2), 2);
-        assertEquals(tableSizeFor(3), 4);
-        assertEquals(tableSizeFor(15), 16);
-        assertEquals(tableSizeFor(16), 16);
-        assertEquals(tableSizeFor(17), 32);
-        int maxSize = 1 << 30;
-        assertEquals(tableSizeFor(maxSize - 1), maxSize);
-        assertEquals(tableSizeFor(maxSize), maxSize);
-        assertEquals(tableSizeFor(maxSize + 1), maxSize);
-        assertEquals(tableSizeFor(Integer.MAX_VALUE), maxSize);
+    // creates a map with size mappings
+    Map<String, String> makeMap(int size) {
+        Map<String, String> map = new HashMap<>();
+        for (int i = 0; i < size; ++i) {
+            putMap(map, i);
+        }
```

Comment:
Replace this loop with a call to `putN`.

Reply:
@stuart-marks done.